client/xdag_config.c

```diff
@@ -0,0 +1,536 @@
+        {
+            free(node->name);
+        }
+        for(i=0; i<node->number_keys; i++)
+        {
+        	xdag_config_free_key(node->keys[i]);
+            if(node->key_names[i] != NULL)
+            {
+                free(node->key_names[i]);
+            }
+        }
+        if(node->keys != NULL)
+        {
+            free(node->keys);
+        }
+
+        if(node->key_names != NULL)
+        {
+            free(node->key_names);
+        }
+        free(node);
+    }
+}
+
+
+static void xdag_config_free(CFG *cfg)
+{
+    if(cfg != NULL)
+    {
+        int i;
+        if(cfg->path != NULL)
+        {
+            free(cfg->path);
+        }
+        for(i=0; i<cfg->number_nodes; i++)
+        {
+        	xdag_config_free_node(cfg->nodes[i]);
+            if(cfg->node_names[i] != NULL)
+            {
+                free(cfg->node_names[i]);
+            }
+        }
+        if(cfg->nodes != NULL)
+        {
+            free(cfg->nodes);
+        }
+        if(cfg->node_names != NULL)
+        {
+            free(cfg->node_names);
+        }
+        free(cfg);
+    }
+}
+
+static NODE *xdag_get_node(CFG *config, const char *node_name)
+{
+    int i;
+    for(i=0; i<config->number_nodes; i++)
+    {
+        if(strkscmp(node_name, config->nodes[i]->name) == 0)
+        {
+            return config->nodes[i];
+        }
+    }
+    return NULL;
+}
+
+
+static KEY *xdag_config_get_key(NODE *node, const char *key_name)
+{
+    int i;
+    for(i=0; i<node->number_keys; i++)
+    {
+        if(strkscmp(key_name, node->keys[i]->name) == 0)
+        {
+            return node->keys[i];
+        }
+    }
+    return NULL;
+}
+
+static const char *xdag_config_get_value(void *cfg, const char *node_name, const char *key, const char *default_value)
+{
+    CFG *c = (CFG *)cfg;
+    NODE *node = xdag_get_node(c, node_name);
+
+    if(node != NULL)
+    {
+        KEY *k = xdag_config_get_key(node, key);
+        if(k != NULL && k->value != NULL)
+        {
+            return k->value;
+        }
+    }
+    return default_value;
+}
+
+
+
+static char *xdag_config_read_node_name(FILE *fp)
+{
+    CFG_CONTENT_TYPE type;
+    long offset = ftell(fp);
+    char *line = NULL;
+
+    while((line = trim(readline(fp))) != NULL)
+    {
+        type = xdag_get_content_type(line);
+        if(type == TYPE_NODE)
+        {
+            int len = (int)strlen(line);
+            char *name = (char *)malloc(len-2+1);
+            memset(name, 0, len-2+1);
+            strncpy(name, line+1, len-2);
+            free(line);
+            return name;
+        }
+        else if(type == TYPE_KEY)
+        {
+            free(line);
+            fseek(fp, offset, SEEK_SET);
+            return strdup("");
+        }
+        else
+        {
+            free(line);
+            offset = ftell(fp);
+            continue;
+        }
+    }
+    return NULL;
+}
+
+
+bool isEmpty(const char *s)
+{
+    return (s == NULL) || (strlen(s) == 0);
```

Comment:
You can do `return (s == NULL) || (*s == 0);`

Reply:
Thank you, I am going to modify it.